docs/developer/api_guidelines.md

```diff
@@ -0,0 +1,438 @@
+float - scientific notation (optional exponent e or E) 
+```
+
+**Example:**
+
+```
+failedAttempt >= 3
+average >= 1.0E2
+responseTimeSec > 0.23
+```
+
+## Filter based on strings
+
+**Operators:**
+
+```
+= != contains, starts-with, ends-with
+```
+
+**Format**:
+
+```
+strings wrapped in single quotes
+special characters escaped with \
+```
+
+**Example:**
+
+```
+projectName = 'myproject'
+```
+
+## Filter based on booleans
+
+**Operators:**
+
+```
+= !=
+```
+
+**Format**:
+
+```
+true false
+```
+
+**Example:**
+
+```
+`active != true`
+```
+
+## Filter based on Date / Time
+
+**Operators:**
+
+```
+= != < <= > >=
+```
+
+**Format**:
+
+```
+ISO 8601 string
+```
+
+**Example:**
+
+```
+lastModified >= '2020-01-01T01:00:00Z
+```
+
+---
+# JSON Format
+
+Payload sent to and returned by API endpoints **should** be represented as JSON objects.  
+The structure of these objects **should** be documented in the swagger API docs.
+When accepting JSON payload from a client, the API endpoint **should** validate if the payload adheres to the documented structure, and shall return an error
+if it does not.
+However, if a client sends an object with an unknown property, the API endpoint **should** just ignore it, and proceed with the execution of the request.
+
+Groups of properties that are mutually exclusive from each other **should** be stored in nested objects, instead of having all properties in a flat JSON object.
+
+*Example:*
+
+*NOT OK*:
+```json
+{
+  "name": "my-credential-config",
+  "remoteURL": "my-url:8080", 
+  "httpsToken": "some-value",
+  "httpsUser": "my-user",
+  "sshKey": "another-value",
+  "sshKeyPassword": "something"
+}
+```
+
+*OK:*
+```json
+{
+  "name": "my-credential-config",
+  "remoteURL": "my-url:8080", 
+  "https": {
+    "token": "some-value",
+    "user": "my-user"
+  },
+  "ssh": {
+    "key": "another-value",
+    "keyPassword": "something"
+  }
+}
+```
+---
+# Asynchronous Operations
+
+For requests that trigger a long running/asynchronous operation, the API endpoint **should** respond with a `202 -Accepted` HTTP response code.
+The returned payload **should** include an identifier for the job triggered by the request. This identifier should enable clients to track the current state of the
+job, or to perform other operations on it (e.g. cancelling, pausing). It **should** be suffixed with `ID`
+
+*Example:*
+
+```
+POST /projects/19340-29304:migrate
+
+Status: 202 - Accepted
+
+Body:
+
+{
+  "jobId": "123-456-789"
+}
+```
+
+After accepting and starting/queueing a long running operation, the server **may** provide the current state of the operation via a `GET` endpoint. Example:
+
+```
+GET /projects/123-456-789:migrate?job-id=123-456-789
+
+200 OK
+{
+  "status": {...}
+}
+```
+---
+# Naming Conventions
+
+## Component Names
+
+Service names **should** be concise and as simple as possible. They **should** be written in *kebap-case*. Also, since the name of a service will also be used
```

Comment:
```suggestion
Service names **should** be concise and as simple as possible. They **should** be written in *kebab-case*. Also, since the name of a service will also be used
```